### PR TITLE
Add from_unixtime_nanos operation

### PR DESCRIPTION
**Pull request by alec-heif**
This addresses the nanos use case of #5046 and also adds support for converting nanos that don't fit in a BIGINT.

Related to #5038

### CODE REVIEW

**Comment by alec-heif** [comment]
I signed + emailed the CLA yesterday but I guess it hasn't been processed yet

**Comment by martint** [comment]
@cla-bot check

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java [hunk 1]
```diff
@@ -265,6 +269,32 @@ public void testFromUnixTimeWithTimeZone()
         assertFunction(format("from_unixtime(7200, '%s')", zoneId), TIMESTAMP_WITH_TIME_ZONE, toTimestampWithTimeZone(expected));
     }
 
+    @Test
+    public void testFromUnixTimeNanos_long()
```
Comment (losipiuk) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java:273]: Do not use undescores in method names

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java [hunk 2]
```diff
@@ -265,6 +269,32 @@ public void testFromUnixTimeWithTimeZone()
...
+    {
+        LocalDateTime dateTime = LocalDateTime.of(2001, 1, 22, 3, 4, 5, 123456789);
```
Comment (losipiuk) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java:275]: It seems that you can use `DateTimeTestingUtils.sqlTimestampOf` to construct expected value.
If not, please extract the method needed to construct expected value.
Preferably the test here would be just a series of:
```
assertFunction(from_unixtime_nanos(....), TIMESTAMP_NANOS, sqlTimestampOf(......)
assertFunction(from_unixtime_nanos(....), TIMESTAMP_NANOS, sqlTimestampOf(......)
```

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java [hunk 3]
```diff
@@ -139,6 +144,28 @@ public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime,
...
+        long epochSeconds = unixTimeNanos / NANOSECONDS_IN_SECOND;
+        long nanosOfSecond = unixTimeNanos % NANOSECONDS_IN_SECOND;
+        long picosOfSecond = 1000L * nanosOfSecond;
+        return DateTimes.longTimestamp(epochSeconds, picosOfSecond);
+    }
+
+    @ScalarFunction("from_unixtime_nanos")
+    @LiteralParameters({"p", "s"})
+    @SqlType("timestamp(9)")
+    public static LongTimestamp fromUnixTimeNanos(@SqlType("decimal(p, s)") Slice arg)
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java:160]: use `unixTimeNanos` instead `arg`

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java [hunk 4]
```diff
@@ -139,6 +144,28 @@ public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime,
...
+    {
+        BigInteger unixTimeNanos = Decimals.decodeUnscaledValue(arg);
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java:162]: This is incorrect because it returns `BigInteger` representing whole "unscaled" value. Not just integral part.
That is for `DECIMAL 123.456` you will get `123456`.
Here we want to look at just integral part (module the fact that we want to add 1 to it if factional part is >= 0.5)

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java [hunk 5]
```diff
@@ -139,6 +144,28 @@ public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime,
         return packDateTimeWithZone(Math.round(unixTime * 1000), zoneId.toStringUtf8());
     }
 
+    @ScalarFunction("from_unixtime_nanos")
+    @SqlType("timestamp(9)")
+    public static LongTimestamp fromUnixTimeNanos(@SqlType(StandardTypes.BIGINT) long unixTimeNanos)
+    {
+        long epochSeconds = unixTimeNanos / NANOSECONDS_IN_SECOND;
+        long nanosOfSecond = unixTimeNanos % NANOSECONDS_IN_SECOND;
+        long picosOfSecond = 1000L * nanosOfSecond;
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java:153]: use `PICOSECONDS_PER_NANOSECOND`

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java [hunk 6]
```diff
@@ -139,6 +144,28 @@ public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime,
...
+
+    @ScalarFunction("from_unixtime_nanos")
+    @LiteralParameters({"p", "s"})
+    @SqlType("timestamp(9)")
+    public static LongTimestamp fromUnixTimeNanos(@SqlType("decimal(p, s)") Slice arg)
+    {
+        BigInteger unixTimeNanos = Decimals.decodeUnscaledValue(arg);
+        long epochSeconds = unixTimeNanos.divide(BigInteger.valueOf(NANOSECONDS_IN_SECOND)).longValue();
+        long nanosOfSecond = unixTimeNanos.mod(BigInteger.valueOf(NANOSECONDS_IN_SECOND)).longValue();
+        long picosOfSecond = 1000L * nanosOfSecond;
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java:165]: Use `Timestamps.PICOSECONDS_PER_NANOSECOND`

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java [hunk 7]
```diff
@@ -77,6 +81,7 @@
     private static final int MILLISECONDS_IN_MINUTE = 60 * MILLISECONDS_IN_SECOND;
     private static final int MILLISECONDS_IN_HOUR = 60 * MILLISECONDS_IN_MINUTE;
     private static final int MILLISECONDS_IN_DAY = 24 * MILLISECONDS_IN_HOUR;
+    private static final long NANOSECONDS_IN_SECOND = 1_000_000_000L;
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java:84]: Use `Timestamps.NANOSECONDS_PER_SECOND`

**Review by losipiuk — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java [hunk 8]
```diff
@@ -139,6 +146,28 @@ public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime,
         return packDateTimeWithZone(Math.round(unixTime * 1000), zoneId.toStringUtf8());
     }
 
+    @ScalarFunction("from_unixtime_nanos")
+    @SqlType("timestamp(9)")
+    public static LongTimestamp fromUnixTimeNanos(@SqlType(StandardTypes.BIGINT) long unixTimeNanos)
+    {
+        long epochSeconds = unixTimeNanos / Timestamps.NANOSECONDS_PER_SECOND;
+        long nanosOfSecond = unixTimeNanos % Timestamps.NANOSECONDS_PER_SECOND;
+        long picosOfSecond = Timestamps.PICOSECONDS_PER_NANOSECOND * nanosOfSecond;
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java:155]:
```suggestion
        long picosOfSecond =  nanosOfSecond * Timestamps.PICOSECONDS_PER_NANOSECOND;
```

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java [hunk 9]
```diff
@@ -265,6 +265,16 @@ public void testFromUnixTimeWithTimeZone()
         assertFunction(format("from_unixtime(7200, '%s')", zoneId), TIMESTAMP_WITH_TIME_ZONE, toTimestampWithTimeZone(expected));
     }
 
+    @Test
+    public void testFromUnixTimeNanos()
```
Comment (losipiuk) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java:269]: put it below `testFromUnixTime`

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java [hunk 10]
```diff
@@ -265,6 +265,16 @@ public void testFromUnixTimeWithTimeZone()
...
+    {
+        assertFunction("from_unixtime_nanos(1234567890123456789)", TIMESTAMP_NANOS, sqlTimestampOf(9, 1234567890L, 123456789));
+        assertFunction("from_unixtime_nanos(1234567890000000000)", TIMESTAMP_NANOS, sqlTimestampOf(9, 1234567890L, 0));
+        assertFunction("from_unixtime_nanos(999999999)", TIMESTAMP_NANOS, sqlTimestampOf(9, 0, 999999999));
+        assertFunction("from_unixtime_nanos(DECIMAL '12345678900123456789')", TIMESTAMP_NANOS, sqlTimestampOf(9, 12345678900L, 123456789));
+        assertFunction("from_unixtime_nanos(DECIMAL '12345678900123456789.999')", TIMESTAMP_NANOS, sqlTimestampOf(9, 12345678900L, 123456789));
```
Comment (losipiuk) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java:275]: Shouldn't we round in such case? @findepi @martint ?

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java [hunk 11]
```diff
@@ -231,8 +232,7 @@ public void testFromUnixTimeWithOffset()
         DateTime expected = new DateTime(dateTime, getDateTimeZone(getTimeZoneKeyForOffset((timeZoneHoursOffset * 60L) + timezoneMinutesOffset)));
         assertFunction("from_unixtime(" + seconds + ", " + timeZoneHoursOffset + ", " + timezoneMinutesOffset + ")", TIMESTAMP_WITH_TIME_ZONE, toTimestampWithTimeZone(expected));
 
-        // test invalid minute offsets
-        assertInvalidFunction("from_unixtime(0, 1, 10000)", INVALID_FUNCTION_ARGUMENT);
+        // test invalid minute offsets assertInvalidFunction("from_unixtime(0, 1, 10000)", INVALID_FUNCTION_ARGUMENT);
```
Comment (losipiuk) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java:235]: revert

File: presto-main/src/main/java/io/prestosql/testing/DateTimeTestingUtils.java [hunk 12]
```diff
@@ -43,6 +45,13 @@ public static SqlTimestamp sqlTimestampOf(
         return sqlTimestampOf(precision, LocalDateTime.of(year, monthOfYear, dayOfMonth, hourOfDay, minuteOfHour, secondOfMinute, millisToNanos(millisOfSecond)));
     }
 
+    public static SqlTimestamp sqlTimestampOf(int precision, long epochSeconds, int epochNanos)
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/testing/DateTimeTestingUtils.java:48]: I do not see a benefit of adding this method vs using `SqlTimestamp.newInstance(precision, epochMicros, picosOfMicro)` directly in test

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java [hunk 13]
```diff
@@ -139,6 +146,28 @@ public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime,
...
+        return DateTimes.longTimestamp(epochSeconds, picosOfSecond);
+    }
+
+    @ScalarFunction("from_unixtime_nanos")
+    @LiteralParameters({"p", "s"})
+    @SqlType("timestamp(9)")
+    public static LongTimestamp fromUnixTimeNanos(@LiteralParameter("s") long numScale, @SqlType("decimal(p, s)") Slice unixTimeNanosDecimal)
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java:162]: You also need a version takes short version of decimal, which is represented as `long`.
Otherwise call like this one `from_unixtime_nanos(DECIMAL '123456789)` will fail.

You can group two implementation in internal static class. See https://github.com/prestosql/presto/blob/2379c52ead0ca9f50d89fd75aa25e885a81118fa/presto-main/src/main/java/io/prestosql/operator/scalar/MathFunctions.java#L140-L166 as an example

**Review by losipiuk — CHANGES_REQUESTED** [review]
Thanks!
Would you be fine also adding documentation to `datetime.rst`?

Also please  update PR message to reference https://github.com/prestosql/presto/issues/5038

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java [hunk 14]
```diff
@@ -139,6 +146,28 @@ public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime,
         return packDateTimeWithZone(Math.round(unixTime * 1000), zoneId.toStringUtf8());
     }
 
+    @ScalarFunction("from_unixtime_nanos")
+    @SqlType("timestamp(9)")
+    public static LongTimestamp fromUnixTimeNanos(@SqlType(StandardTypes.BIGINT) long unixTimeNanos)
+    {
+        long epochSeconds = unixTimeNanos / Timestamps.NANOSECONDS_PER_SECOND;
+        long nanosOfSecond = unixTimeNanos % Timestamps.NANOSECONDS_PER_SECOND;
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java:154]: This needs to use `floorDiv` and `floorMod` to produce correct results when the input is negative. In particular, nanosOfSecond should never be negative

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java (hunk 8 above)
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java:155]: Static import this constant

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java (hunk 13 above)
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java:162]:
```suggestion
    public static LongTimestamp fromUnixTimeNanos(@LiteralParameter("s") long scale, @SqlType("decimal(p, s)") Slice unixTimeNanos)
```

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java [hunk 15]
```diff
@@ -139,6 +146,28 @@ public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime,
...
+    }
+
+    @ScalarFunction("from_unixtime_nanos")
+    @LiteralParameters({"p", "s"})
+    @SqlType("timestamp(9)")
+    public static LongTimestamp fromUnixTimeNanos(@LiteralParameter("s") long numScale, @SqlType("decimal(p, s)") Slice unixTimeNanosDecimal)
+    {
+        BigInteger unixTimeNanos = new BigDecimal(Decimals.toString(unixTimeNanosDecimal, (int) numScale)).toBigInteger();
+        long epochSeconds = unixTimeNanos.divide(BigInteger.valueOf(Timestamps.NANOSECONDS_PER_SECOND)).longValue();
+        long nanosOfSecond = unixTimeNanos.mod(BigInteger.valueOf(Timestamps.NANOSECONDS_PER_SECOND)).longValue();
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java:166]: These need `floorDiv` and `floorMod` semantics, per comment above.

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java (hunk 10 above)
Comment (martint) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java:275]: It should either round or fail, but truncation is inconsistent with datetime semantics in the engine.

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java (hunk 9 above)
Comment (martint) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java:269]: Add some tests with negative unix time

**Review by martint — CHANGES_REQUESTED** [review]

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java (hunk 14 above)
Comment (alec-heif) [reply presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java]: ah interesting, do we need to consider negative unix timestamps as valid? I would prefer to throw when they are encountered.

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java (hunk 10 above)
Comment (alec-heif) [reply presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java]: ok I'll change it to fail

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java (hunk 13 above)
Comment (alec-heif) [reply presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java]: is there somewhere where the implementation details of the short version of decimal are written up? specifically i'm wondering how a non-integer SHORT decimal is represented (e.g. `DECIMAL '123.456'`)

Comment (alec-heif) [reply presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java]: The current branch works for integer SHORT decimals but since we want to now support either rounding or throwing on non-integer DECIMALS I'll now need to worry about this.

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java (hunk 14 above)
Comment (martint) [reply presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java]: I don't see why not. 

From wikipedia (https://en.wikipedia.org/wiki/Unix_time)

> The Unix time number is zero at the Unix epoch, and increases by exactly 86400 per day since the epoch. Thus 2004-09-16T00:00:00Z, 12677 days after the epoch, is represented by the Unix time number 12677 × 86400 = 1095292800. This can be extended backwards from the epoch too, using negative numbers;

Comment (alec-heif) [reply presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java]: 👍 ok will do

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java (hunk 13 above)
Comment (losipiuk) [reply presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java]: > i'm wondering how a non-integer SHORT decimal is represented

I am not sure if it is written up anywhere. We basically keep the digits and the sign in the value. And where the decimal dot should be put is derived from type parameters.

File: presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java [hunk 16]
```diff
@@ -227,6 +227,7 @@ public void testFromUnixTimeNanos()
         assertFunction("from_unixtime_nanos(1234567890123456789)", TIMESTAMP_NANOS, SqlTimestamp.newInstance(9, 1234567890_123456L, 789000));
         assertFunction("from_unixtime_nanos(1234567890000000000)", TIMESTAMP_NANOS, SqlTimestamp.newInstance(9, 1234567890_000000L, 0));
         assertFunction("from_unixtime_nanos(999999999)", TIMESTAMP_NANOS, SqlTimestamp.newInstance(9, 999999, 999000));
+        assertFunction("from_unixtime_nanos(DECIMAL '123456789')", TIMESTAMP_NANOS, SqlTimestamp.newInstance(9, 123456, 789000));
```
Comment (losipiuk) [inline presto-main/src/test/java/io/prestosql/operator/scalar/TestDateTimeFunctions.java:230]: Add test for `DECIMAL '123456789.999'` it would fail now.

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java [hunk 17]
```diff
@@ -148,25 +148,37 @@ public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime,
...
+            long nanosOfSecond = unixTimeNanosInt.mod(BigInteger.valueOf(NANOSECONDS_PER_SECOND)).longValue();
+            long picosOfSecond = nanosOfSecond * PICOSECONDS_PER_NANOSECOND;
+            return DateTimes.longTimestamp(epochSeconds, picosOfSecond);
+        }
+
+        @LiteralParameters({"p", "s"})
+        @SqlType("timestamp(9)")
+        public static LongTimestamp fromShort(@SqlType("decimal(p, s)") long unixTimeNanos)
+        {
+            long epochSeconds = unixTimeNanos / NANOSECONDS_PER_SECOND;
```
Comment (losipiuk) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java:170]: This code would only work for `s==0`. If there is decimal dot in the number we need to take that into account and basically round the value to nearest integer before proceeding with conversion to timestamp.
You can replicate rounding code from https://github.com/prestosql/presto/blob/2379c52ead0ca9f50d89fd75aa25e885a81118fa/presto-main/src/main/java/io/prestosql/operator/scalar/MathFunctions.java#L853-L870

File: presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java [hunk 18]
```diff
@@ -145,6 +153,50 @@ public static long fromUnixTime(@SqlType(StandardTypes.DOUBLE) double unixTime,
...
+
+        @LiteralParameters({"p", "s"})
+        @SqlType("timestamp(9)")
+        public static LongTimestamp fromLong(@LiteralParameter("s") long scale, @SqlType("decimal(p, s)") Slice unixTimeNanos)
+        {
+            BigInteger unixTimeNanosInt = unscaledDecimalToBigInteger(rescale(unixTimeNanos, -(int) scale));
+            long epochSeconds = unixTimeNanosInt.divide(BigInteger.valueOf(NANOSECONDS_PER_SECOND)).longValue();
+            long nanosOfSecond = unixTimeNanosInt.remainder(BigInteger.valueOf(NANOSECONDS_PER_SECOND)).longValue();
+            long picosOfSecond = nanosOfSecond * PICOSECONDS_PER_NANOSECOND;
+            if (picosOfSecond < 0) {
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/DateTimeFunctions.java:169]: Add a comment to indicate that this is simulating floorMod/floorDiv due to lack of direct support in BigInteger

File: presto-docs/src/main/sphinx/functions/datetime.rst [hunk 19]
```diff
@@ -150,6 +150,12 @@ Date and Time Functions
     using ``hours`` and ``minutes`` for the time zone offset. ``unixtime`` is
     the number of seconds since ``1970-01-01 00:00:00`` in ``double`` data type.
 
+.. function:: from_unixtime_nanos(unixtime) -> timestamp(9)
+    :noindex:
```
[the remark this answers is not in the document]
Comment (alec-heif) [reply presto-docs/src/main/sphinx/functions/datetime.rst:154]: I didn't think it made sense to have this as a top-level index item when `from_unixtime` already had one, as it seemed a bit redundant. Happy to remove this though if you think it's sufficiently different it should be a separate item!

[the remark this answers is not in the document]
Comment (martint) [reply presto-docs/src/main/sphinx/functions/datetime.rst:154]: It's a function with different name and result, so I would include it.

**Review by martint — APPROVED** [review]
Looks good. Can you rebase and squash the commits before I merge it?

**Comment by alec-heif** [comment]
> Looks good. Can you rebase and squash the commits before I merge it?

Thanks @martint , should be good to go now!
